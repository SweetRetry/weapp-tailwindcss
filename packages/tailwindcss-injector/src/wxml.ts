import type { Buffer } from 'node:buffer'
import path from 'node:path'
import fs from 'fs-extra'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'

const srcImportTagsMap: Record<string, string[]> = {
  // audio: ['src', 'poster'],
  // video: ['src', 'poster'],
  // image: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxs/01wxs-module.html
  // wxs: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html
  import: ['src'],
  include: ['src'],
}

export interface WxmlDep {
  tagName: string
  start: number
  end: number
  quote: string | null | undefined
  name: string
  value: string
  attrs: Record<string, string>
}

// https://github.com/fb55/htmlparser2/issues/1541
// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer) {
  const ms = new MagicString(wxml.toString())
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}

  const parser = new Parser(
    {
      onopentagname(name) {
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
            }
          }
        }
      },
      onclosetag() {
        currentTagName = ''
        attrs = {}
        importAttrs = undefined
      },
    },
    {
      lowerCaseTags: false,
      xmlMode: true,
    },
  )
  parser.write(
    ms.original,
  )
  parser.end()
  return deps
}

export async function getDepFiles(file: string) {
  const set = new Set<string>()

  async function getDep(filepath: string) {
    if (await fs.exists(filepath)) {
      const dirname = path.dirname(filepath)
      set.add(filepath)
      const wxml = await fs.readFile(filepath, 'utf8')
      const deps = await processWxml(wxml)
      for (const { value } of deps) {
        const p = path.resolve(dirname, value)
        if (!set.has(p)) {
          await getDep(p)
        }
      }
    }
  }

  await getDep(file)

  return set
}
