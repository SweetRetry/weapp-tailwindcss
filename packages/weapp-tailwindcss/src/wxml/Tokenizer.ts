// 针对微信的属性表达式进行切词
// https://github.com/sonofmagic/weapp-tailwindcss/issues/319
// all state
export enum State {
  START,
  TEXT,
  OPEN_BRACE,
  POTENTIAL_CLOSE,
  BRACES_COMPLETE,
}

export class TokenizerStateMachine {
  private state: State
  private buffer: string
  private tokens: string[]

  constructor() {
    this.state = State.START
    this.buffer = ''
    this.tokens = []
  }

  private processChar(char: string) {
    switch (this.state) {
      case State.START:
        if (char === ' ') {
          // Ignore leading spaces
        }
        else if (char === '{') {
          this.state = State.OPEN_BRACE
          this.buffer += char
        }
        else {
          this.state = State.TEXT
          this.buffer += char
        }
        break

      case State.TEXT:
        if (char === ' ') {
          this.tokens.push(this.buffer)
          this.buffer = ''
          this.state = State.START
        }
        else if (char === '{') {
          this.buffer += char
          this.state = State.OPEN_BRACE
        }
        else {
          this.buffer += char
        }
        break

      case State.OPEN_BRACE:
        if (char === '}') {
          this.buffer += char
          this.state = State.POTENTIAL_CLOSE
        }
        else {
          this.buffer += char
        }
        break

      case State.POTENTIAL_CLOSE:
        if (char === '}') {
          this.buffer += char
          this.state = State.BRACES_COMPLETE
        }
        else {
          this.buffer += `}${char}`
          this.state = State.OPEN_BRACE
        }
        break

      case State.BRACES_COMPLETE:
        if (char === ' ') {
          this.tokens.push(this.buffer)
          this.buffer = ''
          this.state = State.START
        }
        else {
          this.buffer += char
          this.state = State.TEXT
        }
        break

      default:
        throw new Error('Unexpected state')
    }
  }

  public run(input: string): string[] {
    for (const char of input) {
      this.processChar(char)
    }
    // Push the last buffer if it's not empty
    if (this.buffer.length > 0) {
      this.tokens.push(this.buffer)
    }
    return this.tokens
  }
}
