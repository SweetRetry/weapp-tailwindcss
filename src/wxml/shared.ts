import { MappingChars2String as dic } from '@/dic'
import { ICommonReplaceOptions } from '@/types'
import { mangleMark } from '@/mangle/expose'

// export function arbitraryValuesReplacer (str: string) {}

export function replaceWxml (
  original: string,
  options: ICommonReplaceOptions | boolean = {
    keepEOL: false,
    mangle: false
  }
) {
  if (typeof options === 'boolean') {
    options = {
      keepEOL: options,
      mangle: false
    }
  }
  let res = original
    .replace(/\[/g, dic['['])
    .replace(/\]/g, dic[']'])
    .replace(/\(/g, dic['('])
    .replace(/\)/g, dic[')'])
    .replace(/#/g, dic['#']) // hex
    .replace(/!/g, dic['!']) // css !important
    .replace(/\//g, dic['/'])
    .replace(/\./g, dic['.'])
    .replace(/:/g, dic[':'])
    // https://github.com/sonofmagic/weapp-tailwindcss-webpack-plugin/issues/8
    .replace(/%/g, dic['%'])
    .replace(/,/g, dic[','])
    .replace(/\\/g, dic['\\'])
    .replace(/'/g, dic["'"])
    .replace(/"/g, dic['"'])
    .replace(/\*/g, dic['*'])
    .replace(/&/g, dic['&'])
    .replace(/@/g, dic['@'])
    .replace(/{/g, dic['{'])
    .replace(/}/g, dic['}'])
  if (!options.keepEOL) {
    res = res
      // 去除无用换行符和空格
      .replace(/[\r\n]+/g, '')
  }
  if (options.mangle) {
    res = mangleMark(res, options.classGenerator)
  }
  return res
}
