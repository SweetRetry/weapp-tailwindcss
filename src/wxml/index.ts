import { parse, render } from 'wxml-ast'
import { parseExpression } from '@babel/parser'
import traverse from '@babel/traverse'
import generate from '@babel/generator'
import { replaceWxml } from './shared'

export { replaceWxml }

// #region internal
const variableRegExp = /{{([^{}]*)}}/g

function variableMatch(original: string) {
  return variableRegExp.exec(original)
}
// #endregion

export function generateCode(match: string) {
  const ast = parseExpression(match)

  traverse(ast, {
    StringLiteral(path) {
      path.node.value = replaceWxml(path.node.value)
    },
    noScope: true
  })

  const { code } = generate(ast, {
    compact: true,
    minified: true
  })
  return code
}

export interface RawSource {
  start: number
  end: number
  raw: string
  // '' 直接 remove {{}}
  source?: string
}

export function templeteReplacer(original: string) {
  let match = variableMatch(original)
  const sources: RawSource[] = []

  while (match !== null) {
    // 过滤空字符串
    // if (match[1].trim().length) {
    sources.push({
      start: match.index,
      end: variableRegExp.lastIndex,
      raw: match[1]
    })

    match = variableMatch(original)
  }

  if (sources.length) {
    const resultArray: string[] = []
    let p = 0
    for (let i = 0; i < sources.length; i++) {
      const m = sources[i]
      // 匹配前值
      resultArray.push(replaceWxml(original.slice(p, m.start), true))
      p = m.start
      // 匹配后值
      if (m.raw.trim().length) {
        const code = generateCode(m.raw)
        m.source = `{{${code}}}`
      } else {
        m.source = ''
      }

      resultArray.push(m.source)
      p = m.end
      // 匹配最终尾部值
      if (i === sources.length - 1) {
        resultArray.push(replaceWxml(original.slice(m.end), true))
      }
    }

    return resultArray.filter((x) => x).join('')
  } else {
    return replaceWxml(original)
  }
}

export async function templeteHandler(rawSource: string) {
  const parsed = await parse(rawSource, {
    elementCB: (node) => {
      if (node.attribs.class) {
        node.attribs.class = templeteReplacer(node.attribs.class)
      }
    }
  })
  return render(parsed)
  // wxml.traverse(parsed, (node, parent) => {
  //   if (node.type === wxml.NODE_TYPES.ELEMENT) {
  //     // @ts-ignore
  //     if (node.attributes.class) {
  //       // @ts-ignore
  //       node.attributes.class = templeteReplacer(node.attributes.class)
  //     }
  //   }
  // })
  // return wxml.serialize(parsed)
}
