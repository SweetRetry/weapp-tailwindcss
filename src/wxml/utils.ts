import * as t from '@babel/types'
import { replaceWxml } from './shared'
import { parseExpression, traverse, generate } from '@/babel'
import { variableMatch, variableRegExp, templateClassExactRegexp, tagWithEitherClassAndHoverClassRegexp, makeCustomAttributes } from '@/reg'
import { defu } from '@/utils'
import type { RawSource, ICommonReplaceOptions, ITempleteHandlerOptions } from '@/types'

export function generateCode(match: string, options: ICommonReplaceOptions = {}) {
  const ast = parseExpression(match)

  traverse(ast, {
    StringLiteral(path) {
      // [g['人生']==='你好啊'?'highlight':'']
      if (t.isMemberExpression(path.parent)) {
        return
      }
      // parentPath maybe null
      // ['td',[(g.type==='你好啊')?'highlight':'']]
      if (t.isBinaryExpression(path.parent) && t.isConditionalExpression(path.parentPath?.parent)) {
        return
      }

      path.node.value = replaceWxml(path.node.value, options)
    },
    noScope: true
  })
  // @ts-ignore
  const { code } = generate(ast, {
    compact: true,
    minified: true,
    jsescOption: {
      quotes: 'single',
      minimal: true
    }
  })

  return code
}

export function extractSource(original: string) {
  let match = variableMatch(original)
  const sources: RawSource[] = []

  while (match !== null) {
    // 过滤空字符串
    // if (match[1].trim().length) {
    const start = match.index
    const end = variableRegExp.lastIndex
    sources.push({
      start,
      end,
      raw: match[1],
      prevConcatenated: !/\s/.test(original[start - 1]),
      nextConcatenated: !/\s/.test(original[end])
    })

    match = variableMatch(original)
  }
  return sources
}

export function templeteReplacer(original: string, options: ICommonReplaceOptions = {}) {
  const sources = extractSource(original)

  if (sources.length > 0) {
    const resultArray: string[] = []
    let p = 0
    for (let i = 0; i < sources.length; i++) {
      const m = sources[i]
      const before = original.slice(p, m.start)

      // 匹配前值
      resultArray.push(
        replaceWxml(before, {
          keepEOL: true,
          escapeMap: options.escapeMap
        })
      )
      p = m.start
      // 匹配后值
      if (m.raw.trim().length > 0) {
        const code = generateCode(m.raw, options)
        const source = `{{${code}}}`
        m.source = source
      } else {
        m.source = ''
      }

      resultArray.push(m.source)
      p = m.end
      // 匹配最终尾部值
      if (i === sources.length - 1) {
        const after = original.slice(m.end)
        resultArray.push(
          replaceWxml(after, {
            keepEOL: true,
            escapeMap: options.escapeMap
          })
        )
      }
    }

    return resultArray
      .filter(Boolean)
      .join('')
      .trim()
  } else {
    return replaceWxml(original, {
      keepEOL: false,
      escapeMap: options.escapeMap
    })
  }
}

export function templeteHandler(rawSource: string, options: ICommonReplaceOptions = {}) {
  return rawSource.replace(tagWithEitherClassAndHoverClassRegexp, (m0) => {
    return m0.replace(templateClassExactRegexp, (m1, className) => {
      return m1.replace(className, templeteReplacer(className, options))
    })
  })
}

export function customTempleteHandler(rawSource: string, options: ITempleteHandlerOptions = {}) {
  let source = templeteHandler(rawSource, options)
  const regexps = makeCustomAttributes(options.customAttributesEntities)
  if (regexps) {
    if (Array.isArray(regexps)) {
      for (const regexp of regexps) {
        source = source.replace(regexp.tagRegexp, (m0) => {
          return m0.replace(regexp.attrRegexp, (m1, className) => {
            return m1.replace(className, templeteReplacer(className, options))
          })
        })
      }
    }

    return source
  } else {
    return source
  }
}

export function createTempleteHandler(options: ITempleteHandlerOptions = {}) {
  return (rawSource: string, opt: ITempleteHandlerOptions = {}) => {
    return customTempleteHandler(rawSource, defu(opt, options))
  }
}
