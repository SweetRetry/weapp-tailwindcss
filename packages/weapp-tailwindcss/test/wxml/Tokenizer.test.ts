import { TokenizerStateMachine } from '@/wxml/Tokenizer'

describe('tokenizerStateMachine', () => {
  it('should tokenize a string with spaces and text correctly', () => {
    const inputString = '2xl:text-xs rd-tag-{{type}}-{{theme}} {{prefix}}-btn'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['2xl:text-xs', 'rd-tag-{{type}}-{{theme}}', '{{prefix}}-btn'])
  })

  it('should handle leading and trailing spaces', () => {
    const inputString = '  2xl:text-xs  '
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['2xl:text-xs'])
  })

  it('should handle multiple spaces between tokens', () => {
    const inputString = '2xl:text-xs   rd-tag-{{type}}  '
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['2xl:text-xs', 'rd-tag-{{type}}'])
  })

  it('should handle empty string', () => {
    const inputString = ''
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual([])
  })

  it('should handle strings with only spaces', () => {
    const inputString = '     '
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual([])
  })

  it('should handle single token without spaces', () => {
    const inputString = '2xl:text-xs'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['2xl:text-xs'])
  })

  it('should handle multiple tokens without spaces', () => {
    const inputString = '2xl:text-xsrd-tag-{{type}}'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['2xl:text-xsrd-tag-{{type}}'])
  })

  it('should handle nested braces correctly', () => {
    const inputString = 'rd-tag-{{type}}-{{nested-{{inner}}}}'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['rd-tag-{{type}}-{{nested-{{inner}}}}'])
  })

  it('should handle braces with spaces inside', () => {
    const inputString = 'class-{{ some value }}-modifier'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['class-{{ some value }}-modifier'])
  })

  it('should handle braces with JS expressions inside', () => {
    const inputString = 'result-{{ 2 + 2 }}-final'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['result-{{ 2 + 2 }}-final'])
  })

  it('should handle multiple braces with complex expressions', () => {
    const inputString = 'prefix-{{ someValue }}-mid-{{ another + value }}-suffix'
    const tokenizer = new TokenizerStateMachine()
    const result = tokenizer.run(inputString)
    expect(result).toEqual(['prefix-{{ someValue }}-mid-{{ another + value }}-suffix'])
  })
})
