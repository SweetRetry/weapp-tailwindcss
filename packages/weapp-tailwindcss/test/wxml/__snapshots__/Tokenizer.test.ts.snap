// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tokenizer > should tokenize a string with spaces and text correctly 1`] = `
[
  {
    "end": 11,
    "expressions": [],
    "start": 0,
    "value": "2xl:text-xs",
  },
  {
    "end": 37,
    "expressions": [
      {
        "end": 27,
        "start": 19,
        "value": "{{type}}",
      },
      {
        "end": 37,
        "start": 28,
        "value": "{{theme}}",
      },
    ],
    "start": 12,
    "value": "rd-tag-{{type}}-{{theme}}",
  },
  {
    "end": 52,
    "expressions": [
      {
        "end": 48,
        "start": 38,
        "value": "{{prefix}}",
      },
    ],
    "start": 38,
    "value": "{{prefix}}-btn",
  },
]
`;
